### PR TITLE
Return bootstrap interface object, not a promise

### PR DESCRIPTION
**Pull request by JoeChapman**
### This will be a breaking change

- make interface chaninable by returning bootstrap from interface methods
- remove promise so no more `then`
- apply error middleware in the start method so middleware applied with `bootstrap.use` will be applied before error md.
- move some set up to functions outside of module definition
- update readme documentation
- also, fixed lint error in lib/router

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,66 +1,100 @@
...
+    translate: i18n.translate.bind(i18n),
+  }));
+
+  app.use(hof.middleware.errors({
+    translate: i18n.translate.bind(i18n),
+    debug: config.env === 'development'
+  }));
+};
+
+const getI18n = (config) =>
```
Comment (joefitter) [inline index.js]: parens

Comment (JoeChapman) [inline index.js]: glans

File: index.js [hunk 2]
```diff
@@ -1,66 +1,100 @@
...
+  hof.i18n({
+    path: path.resolve(config.caller, config.translations) + '/__lng__/__ns__.json'
```
Comment (joefitter) [inline index.js]: Can we add a check here to test file exists? if not then fallback to hof-template-partials?

File: index.js [hunk 3]
```diff
@@ -1,66 +1,100 @@
...
+      applyErrorMiddlewares(config, getI18n(config));
+
+      bootstrap.server.listen(config.port, config.port);
+      return bootstrap;
     },
 
-    stop: () => {
-      bootstrap.server.close();
+    stop: (callback) => {
+      const interval = setInterval(() => {
```
Comment (joefitter) [inline index.js]: Why is this needed? If we really need to defer can we use `_.defer`

File: index.js (hunk 1 above)
Comment (joefitter) [inline index.js]: Also - does this need to be in a function if its just called once to set up i18n below?

File: index.js [hunk 4]
```diff
@@ -1,66 +1,100 @@
...
+      if (callback) {
+        callback(bootstrap.server);
+      }
+      return bootstrap;
     }
 
   };
 
   const config = getConfig(options);
+  const i18n = getI18n(config);
```
Comment (joefitter) [inline index.js]: init here?

File: index.js (hunk 2 above)
Comment (JoeChapman) [inline index.js]: coming in a second PR

File: index.js (hunk 3 above)
Comment (JoeChapman) [inline index.js]: possibly

File: index.js (hunk 4 above)
Comment (JoeChapman) [inline index.js]: Can you elaborate?

File: index.js (hunk 1 above)
Comment (JoeChapman) [inline index.js]: using a function means we can manage variables more easily. Organising groups into functions avoids us relying on outer scopes and stuff.

Comment (joefitter) [inline index.js]: ES6 `const` and `let` are block scoped not function scoped, I may be wrong but this appears to be doing exactly the same as if it were inline, and as it is only called once it is adding complexity (IMO)

File: index.js (hunk 4 above)
Comment (joefitter) [inline index.js]: Sorry, I just meant the same as above - can't we init inline?

File: index.js (hunk 1 above)
Comment (JoeChapman) [inline index.js]: IMO, it is easier to manage because we use it in the `start` function; https://github.com/UKHomeOffice/hof-bootstrap/pull/46/files#diff-168726dbe96b3ce427e7fedce31bb0bcR65, otherwise we'd need to pass it into `start`, which becomes very tricky when the user calls `start`

Comment (joefitter) [inline index.js]: But then aren't we init-ing i18n twice?

Comment (joefitter) [inline index.js]: you _could_ or we could simply init it once on calling the bootstrap function an store a reference to it, perhaps on the bootstrap object, to be used in later applied middleware, such as these?

Comment (joefitter) [inline index.js]: Also, do we need to pass config to start? I think that convolutes it somewhat, we are calling bootstrap with config, then calling start with potentially different config. Unless there is a use case I am missing?

Comment (JoeChapman) [inline index.js]: I believe that if a developer wanted to defer start, they might also want to defer passing port, host and/or protocol options

**Comment by JoeChapman** [comment]
Can we rebase these commits please

**Comment by joefitter** [comment]
No Joe, where have you been?

**Comment by joefitter** [comment]
squash and merge 💃

**Comment by JoeChapman** [comment]
squash and merge is what I meant. I have been on holiday. It was nice, thanks for asking

**Comment by JoeChapman** [comment]
Before we merge, is the readme update date?

**Comment by joefitter** [comment]
no Joe, that's your domain

**Comment by JoeChapman** [comment]
Please don't merge this with incorrect docs, it's like merging a commit with failing tests
